### PR TITLE
fix: force active membership if the user has active subs

### PR DESCRIPTION
**Pull request by dkoo**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This is a third attempt at addressing the problem described in #3060 and #3050, where if a user has more than one subscription and not all are active, any of the user's memberships tied to the subscriptions' products could be incorrectly expired at any time.

This implements two new checks at runtime:

1. When checking the user membership's status
2. When checking whether the user has access to a restricted post

The first should run only for memberships with non-active statuses, and only if the membership's plan is tied to subscription product ownership. If it is and the user does have an active subscription with the required subscription product(s), it will update the membership to `active` status and relink the membership to the active subscription. 

The second check is a fallback of sorts at the point where the user is trying to view restricted content. It checks if the membership requires any subscription products and allows access as long as the user has an active subscription with one of those products. Since the membership updates from the first check could take longer than it takes for the restricted content to load in the browser, this second check should avoid a race condition the first time any affected users try to view restricted content, but maybe it's not needed since the first check should permanently fix the membership status.

The major piece that was missing from #3050 was `WC_Memberships_User_Membership::unschedule_expiration_events()`—without calling this, updating an expired membership to `active` will result in an infinite loop of active/expire actions that will severely impact site performance. If we call that method before updating the status to `active`, the check should run exactly once for any impacted memberships, so in theory this PR should have no negative effect on site performance.

### How to test the changes in this Pull Request:

1. Set up a membership plan that requires ownership of a subscription product(s) and set it to restrict content (e.g. of a post category).
3. As a reader, purchase the required subscription and confirm that you can view the restricted content.
4. As an admin, manually update the reader subscription to "Expired" status.
5. As the reader, confirm that you can no longer view the restricted content.
6. Still as the reader, purchase another subscription to the required product.
7. On `trunk`, observe that you still can't view restricted content despite owning an active subscription with the required product.
8. As an admin, observe that the reader's membership still shows "Expired" status and is tied to the reader's expired subscription, instead of the newer active one.
9. Check out this branch and refresh the membership either in the admin or the restricted content as the reader.
10. Confirm that the reader can now view restricted content, and that the membership is updated to "Active" status, relinked to the active subscription, and that it doesn't continually try to expire and reactivate itself in an infinite loop that eventually crashes your entire site.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207858308890295

### CODE REVIEW

File: includes/plugins/wc-memberships/class-memberships.php [hunk 1]
```diff
@@ -78,6 +80,13 @@ public static function init() {
 		include __DIR__ . '/class-metering.php';
 	}
 
+	/**
+	 * Check if Memberships is available.
+	 */
+	public static function has_memberships() {
```
Comment (adekbadek) [inline includes/plugins/wc-memberships/class-memberships.php]: Nit: the method name suggests to me that it checks if an entity (site/user) has "memberships", not whether a plugin is active.

File: includes/reader-revenue/woocommerce/class-woocommerce-connection.php [hunk 2]
```diff
@@ -160,20 +195,18 @@ public static function get_last_successful_order( $customer ) {
...
+
 		// Prioritize any currently active subscriptions.
-		$user_subscriptions = \wcs_get_users_subscriptions( $customer->get_id() );
-		if ( ! empty( $user_subscriptions ) ) {
-			foreach ( $user_subscriptions as $subscription ) {
-				if ( $subscription->has_status( self::ACTIVE_SUBSCRIPTION_STATUSES ) ) {
-					return $subscription;
-				}
-			}
+		$active_subscriptions = self::get_active_subscriptions_for_user( $user_id() );
```
Comment (adekbadek) [inline includes/reader-revenue/woocommerce/class-woocommerce-connection.php]: 🤔 
```suggestion
		$active_subscriptions = self::get_active_subscriptions_for_user( $user_id );
```

File: includes/reader-revenue/woocommerce/class-woocommerce-connection.php [hunk 3]
```diff
@@ -160,20 +195,18 @@ public static function get_last_successful_order( $customer ) {
...
+		if ( ! empty( $active_subscriptions ) ) {
+			return reset( $active_subscriptions );
 		}
 
 		// If no active subscriptions, get the most recent completed order.
 		// See https://github.com/woocommerce/woocommerce/wiki/wc_get_orders-and-WC_Order_Query for query args.
 		$args = [
-			'customer_id' => $customer->get_id(),
+			'customer_id' => $user_id(),
```
Comment (adekbadek) [inline includes/reader-revenue/woocommerce/class-woocommerce-connection.php]:
```suggestion
			'customer_id' => $user_id,
```

**Review by adekbadek — CHANGES_REQUESTED** [review]
Works as described! There are two typos to fix in `get_last_successful_order` method.

The testing instructions mention comparing to the `release` branch. Should this PR be a hotfix?

**Comment by dkoo** [comment]
> The testing instructions mention comparing to the `release` branch. Should this PR be a hotfix?

It was originally a hotfix, but I walked that back since it's a pretty major change that warrants more careful QA, I think. I updated the testing instructions.

File: includes/reader-revenue/woocommerce/class-woocommerce-connection.php (hunk 3 above)
Comment (dkoo) [reply includes/reader-revenue/woocommerce/class-woocommerce-connection.php]: Whoops! Fixed in 01d8e275

File: includes/reader-revenue/woocommerce/class-woocommerce-connection.php (hunk 2 above)
Comment (dkoo) [reply includes/reader-revenue/woocommerce/class-woocommerce-connection.php]: Whoops! Fixed in 01d8e275

File: includes/plugins/wc-memberships/class-memberships.php (hunk 1 above)
Comment (dkoo) [reply includes/plugins/wc-memberships/class-memberships.php]: Good point! 01d8e275 updates this to `is_active()` instead.

**Review by adekbadek — APPROVED** [review]

File: includes/reader-revenue/woocommerce/class-woocommerce-connection.php [hunk 4]
```diff
@@ -160,20 +195,18 @@ public static function get_last_successful_order( $customer ) {
...
+
 		// Prioritize any currently active subscriptions.
-		$user_subscriptions = \wcs_get_users_subscriptions( $customer->get_id() );
-		if ( ! empty( $user_subscriptions ) ) {
-			foreach ( $user_subscriptions as $subscription ) {
-				if ( $subscription->has_status( self::ACTIVE_SUBSCRIPTION_STATUSES ) ) {
-					return $subscription;
-				}
-			}
+		$active_subscriptions = self::get_active_subscriptions_for_user( $user_id );
```
Comment (miguelpeixe) [inline includes/reader-revenue/woocommerce/class-woocommerce-connection.php:201]: This method returns an array of integers, which breaks the expected return of `get_last_successful_order()`. This causes the `get_contact_from_customer()` to throw a fatal error by trying to `->get_id()` from the integer and breaks data event handlers that rely on that method.

Comment (dkoo) [reply includes/reader-revenue/woocommerce/class-woocommerce-connection.php:201]: Thanks for catching! I'll have a look and push an "alpha hotfix" for it

Comment (dkoo) [reply includes/reader-revenue/woocommerce/class-woocommerce-connection.php:201]: PR here: https://github.com/Automattic/newspack-plugin/pull/3299

**Pull request merged**